Rubi ruleset using MatchPy (Python 3.6 only)

Upabjojr opened the pull request.

arihantparsoya committed:
Implemented rubi_integrate() function

arihantparsoya committed:
added utility functions

arihantparsoya committed:
added test_rubi

arihantparsoya committed:
Updates with_renamed_vars() function in Constraint

arihantparsoya committed:
added test_rubi.py

arihantparsoya committed:
added more utility functions

arihantparsoya committed:
updated ConstantSymbol()

arihantparsoya committed:
added more rules in patterns.py file

arihantparsoya committed:
removed excess code

Abdullahjavednesar committed:
Added few more test sets

Abdullahjavednesar committed:
Added description for test sets in first line

Abdullahjavednesar committed:
few more utility functions added

Abdullahjavednesar committed:
removed sqrt

arihantparsoya committed:
added more patterns

Abdullahjavednesar committed:
added more definitions and tests

Abdullahjavednesar committed:
Utility functions and tests ready

Abdullahjavednesar committed:
test func name changed

arihantparsoya committed:
Merge pull request #4 from Abdullahjavednesar/ajn_rubi

Added few more test sets

arihantparsoya committed:
improved parser and added few more tests

arihantparsoya committed:
added tests from rubi test suit, added Equal()

arihantparsoya committed:
implemented Set(), WIth(), LessEqual()

arihantparsoya committed:
implemented Less()

arihantparsoya committed:
added more utility functions

arihantparsoya committed:
added all Algebraic rules 1.2 patterns

arihantparsoya committed:
removed `rubi_object()` from `rubi_integrate()` to prevent loading it multiple times

arihantparsoya committed:
typo

Abdullahjavednesar committed:
added With and Rt

arihantparsoya committed:
added IntLinearcQ()

arihantparsoya committed:
added utility functions

arihantparsoya committed:
implemented set1 functions

Abdullahjavednesar committed:
added set 2

Abdullahjavednesar committed:
added set2

Abdullahjavednesar committed:
added tests

arihantparsoya committed:
Merge pull request #6 from Abdullahjavednesar/ajn_rubi2

added With and Rt

arihantparsoya committed:
added utility functions

arihantparsoya committed:
removed unnecessary files

arihantparsoya committed:
added few utility functions and their tests. Added support for is_Add functions in Int()

Abdullahjavednesar committed:
changed tests

Abdullahjavednesar committed:
resolved conflicts

arihantparsoya committed:
removed Int() from utility functions file

arihantparsoya committed:
Merge pull request #7 from Abdullahjavednesar/ajn_rubi3

changed tests

arihantparsoya committed:
added more utility functions. Added rubi_test_suit.py file

arihantparsoya committed:
implemented CoefficientList and ReplaceAll

arihantparsoya committed:
implemented ExpandLinearProduct and SimplifyTerm

arihantparsoya committed:
implemented ExpandIntegrand

arihantparsoya committed:
added documentation for matchQ

Abdullahjavednesar committed:
BinomialParts

Abdullahjavednesar committed:
added Exponent

Abdullahjavednesar committed:
added few more examples

arihantparsoya committed:
implemented ExpandAlgebraicFunction

arihantparsoya committed:
implemented AlgebraicFunctionQ and ExpandCleanup

Abdullahjavednesar committed:
simpify numbers in tests

Abdullahjavednesar committed:
resolved conflicts

arihantparsoya committed:
added more utility functions. Sympified the tests

Abdullahjavednesar committed:
added 2nd and 3rd test suits

arihantparsoya committed:
Merge pull request #8 from Abdullahjavednesar/ajn_rubi4

Added few more definitions

arihantparsoya committed:
fixed First and Rest

arihantparsoya committed:
removed FreeQ() from cons()

arihantparsoya committed:
removed `try` from constraints

arihantparsoya committed:
implemented expandintegrand

arihantparsoya committed:
implemented PerfectPowerTest and FactorSquareFreeList

arihantparsoya committed:
implemented more utility functions

SquareFreeFactorTest
RationalFunctionQ
RationalFunctionFactors
NonrationalFunctionFactors
Reverse
RationalFunctionExponents

arihantparsoya committed:
added more utility functions

QuotientOfLinearsQ
QuotientOfLinearsParts
QuotientOfLinearsP
NonalgebraicFunctionFactors
AlgebraicFunctionFactors
PolyGCD
PolynomialGCD

arihantparsoya committed:
implemented matchpy2sympy function

arihantparsoya committed:
added utility functions

NonabsurdNumberFactors
AbsurdNumberFactors
AbsurdNumberQ
Sort
Flatten

arihantparsoya committed:
commit rubi_test_suit.py

arihantparsoya committed:
added documentation for test_suit function

arihantparsoya committed:
added sumsimplerQ

Abdullahjavednesar committed:
set 7

Abdullahjavednesar committed:
updated

arihantparsoya committed:
added SubstForFractionalPower and SubstForInverseFunction

arihantparsoya committed:
added utility functions

FractionalPowerSubexpressionQ
FractionalPowerOfSquareQ

arihantparsoya committed:
added utility functions

FactorNumericGcd

arihantparsoya committed:
fixed test

arihantparsoya committed:
added utility functions

TrigSimplify
TrigSimplifyQ
MergeFactors
MergeFactor
MergeableFactorQ
FactorAbsurdNumber
FactorInteger
Drop

arihantparsoya committed:
added utility functions

FactorOrder
Order
Smallest

arihantparsoya committed:
implemented new functions

OrderedQ
MinimumDegree

arihantparsoya committed:
added utility functions

NonpositiveFactors
Sign
PositiveFactors

arihantparsoya committed:
added utility functions

PolynomialInQ

arihantparsoya committed:
added utility function

ExponentIn

arihantparsoya committed:
added PolynomialInSubst

arihantparsoya committed:
implemented Distrib

arihantparsoya committed:
added DistributeDegree

arihantparsoya committed:
added all 1.2 patterns

arihantparsoya committed:
added try in constraint.py to avoid `zoo` errors

arihantparsoya committed:
added FunctionOfPower

arihantparsoya committed:
added DivideDegreesOfFactors

Abdullahjavednesar committed:
added tests

Abdullahjavednesar committed:
resolved

arihantparsoya committed:
added MonomialFactor

Abdullahjavednesar committed:
removed duplicates

arihantparsoya committed:
fixed typo

arihantparsoya committed:
added FunctionOfInverseLinear

arihantparsoya committed:
added utility functions

PureFunctionOfSinhQ
PureFunctionOfCoshQ
PureFunctionOfTanhQ
PureFunctionOfCothQ

arihantparsoya committed:
added utility functions

IntegerQuotientQ
OddQuotientQ
EvenQuotientQ
FindTrigFactor
FunctionOfSinhQ

arihantparsoya committed:
added FunctionOfCoshQ

arihantparsoya committed:
added FunctionOfTanhQ

arihantparsoya committed:
added FunctionOfTanhWeight

arihantparsoya committed:
added FunctionOfHyperbolicQ

arihantparsoya committed:
added SmartDenominator and SmartNumerator

arihantparsoya committed:
added SubstForAux

arihantparsoya committed:
added SubstForTrig

arihantparsoya committed:
added SubstForHyperbolic

arihantparsoya committed:
added InertTrigFreeQ

arihantparsoya committed:
added SubstForFractionalPowerOfLinear

arihantparsoya committed:
added InverseFunctionOfLinear

arihantparsoya committed:
added InertTrigQ

arihantparsoya committed:
removed utility_functions from __init__.py

arihantparsoya committed:
added matchpy2sympy in constraints

arihantparsoya committed:
added sympy paths

arihantparsoya committed:
rubi returns NotImplementedError if not integrated.

arihantparsoya committed:
modified Hypergeometrif2F1()

arihantparsoya committed:
removed test_rubi.py

arihantparsoya committed:
added PowerOfInertTrigSumQ and DeactivateTrigAux

arihantparsoya committed:
added signal in rubi test suit

arihantparsoya committed:
fixed a bug in rubi.py and added some utility functions

arihantparsoya committed:
modified Apart

arihantparsoya committed:
modified rubi_integrate() definiton

arihantparsoya committed:
modified MemberQ and completed PiecewiseLinearQ

arihantparsoya committed:
added KnownTrigIntegrandQ

Abdullahjavednesar committed:
set 10-11

Abdullahjavednesar committed:
resolved

arihantparsoya committed:
Merge pull request #9 from Abdullahjavednesar/ajn_rubi5

added till set 7

arihantparsoya committed:
minor changes in matchpy2sympy

arihantparsoya committed:
added showsteps

arihantparsoya committed:
fixed some utility functions

arihantparsoya committed:
fixed AbsorbMinusSign

arihantparsoya committed:
added TryPureTanSubst, TryTanhSubst, TryPureTanhSubst

arihantparsoya committed:
added TrinomialQ

arihantparsoya committed:
added GeneralizedTrinomialQ and GeneralizedBinomialQ

arihantparsoya committed:
added FractionalPowerOfQuotientOfLinears

arihantparsoya committed:
added SubstForFractionalPowerQ

arihantparsoya committed:
added AbsurdNumberGCD

Abdullahjavednesar committed:
fixed minor bugs

Abdullahjavednesar committed:
added a test case

Abdullahjavednesar committed:
resolved

arihantparsoya committed:
Merge pull request #10 from Abdullahjavednesar/ajn_rubi7

Fixed minor bugs

arihantparsoya committed:
Merge branch 'rubi4' of https://github.com/parsoyaarihant/sympy into rubi4

arihantparsoya committed:
fixed few tests

Abdullahjavednesar committed:
fixed MergeMonomials

Abdullahjavednesar committed:
cleaned

arihantparsoya committed:
Merge pull request #11 from Abdullahjavednesar/ajn_rubi8

Fixed Mergemonomials

arihantparsoya committed:
added Map2, ConstantFactor, SameQ, CommonFactors

arihantparsoya committed:
added tests for ExpandIntegrand

arihantparsoya committed:
fixed TryPureTanhSubst test

arihantparsoya committed:
added RationalFunctionExpand

arihantparsoya committed:
added MostMainFactorPosition and CommonFactors and ConstantFactor

arihantparsoya committed:
removed unwanted lines

arihantparsoya committed:
added FunctionOfLinear

arihantparsoya committed:
fixed NormalizeIntegrand

arihantparsoya committed:
added FunctionOfExponential

asmeurer commented on sympy/rubi/sympy2matchpy.py:
```diff
@@ -0,0 +1,38 @@
+Contains Sympy to MatchPy converter function.
+'''
+
+import matchpy
```
You'll need to move matchpy imports inline so that they are only imported in Python 3.6.

arihantparsoya committed:
Improved expandintegrand() and added tests

Upabjojr commented:
@parsoyaarihant maybe it's better that you open this PR.

parsoyaarihant commented:
Sure.

arihantparsoya committed:
fixed imports

arihantparsoya committed:
fixed imports

arihantparsoya committed:
added one_identity to Pow()

parsoyaarihant commented:
How should I arrange the test suit? Should they be divided into type of expressions? Currently they are all arranged into a single list.

For now, I am removing the test suit for algebriac rules and adding some tests for `rubi_integrate()`. Once the test suit pass, I will add them.

arihantparsoya committed:
removed test suit

Upabjojr commented:
> How should I arrange the test suit? Should they be divided into type of expressions? Currently they are all arranged into a single list.

The main question is how to make MatchPy work in the tests. I believe you have to edit `.travis.yml`.

arihantparsoya committed:
added test_rubi.py

Upabjojr commented:
@parsoyaarihant 

Try to import `matchpy` as:

```python
from sympy.external import import_module
matchpy = import_module("matchpy")

if matchpy is None:
    ... matchpy is not installed ...
```

arihantparsoya committed:
added new patterns in ExpandIntegrand

arihantparsoya committed:
added test_rubi_integrate.py

arihantparsoya committed:
removed file

arihantparsoya committed:
fixed imports

arihantparsoya committed:
test_travis test

fixed travis

.travis.yml edit

Update .travis.yml

travis

added matchpy imports

Upabjojr commented:
@parsoyaarihant need to merge `.travis.yml` into your branch, there are changes in the master branch.

parsoyaarihant commented:
Yes, I have done that just now.

asmeurer commented on .travis.yml:
```diff
@@ -15,6 +15,7 @@ python:
   - 3.3
   - 3.4
   - 3.5
+  - 3.6
```
I think there is a bug in GitHub here. These changes are already in master.

parsoyaarihant replied:
The test doesn't seem to be running for this PR. Is that Travis problem?

jksuom commented on sympy/core/compatibility.py:
```diff
@@ -857,6 +857,6 @@ def cache_clear():
 ### End of backported lru_cache
 
-if sys.version_info[:2] >= (3, 3):
+if sys.version_info[:2] >= (3, 6):
```
This is not the place for the matchpy version test. It could be in `rubi/__init__.py` to control importing. (The import of lru_cache should not be changed.)

parsoyaarihant replied:
Thanks!

arihantparsoya committed:
fixed compatibility

arihantparsoya committed:
passed test_setup.py

arihantparsoya committed:
added more patterns in expandintegrand

fixed imports

fixed code quality

arihantparsoya committed:
Completed ExpandIntegrand

parsoyaarihant commented:
Python 3.6 tests are passing

arihantparsoya committed:
added more tests in test_rubi.py

Upabjojr commented:
OK, we need to make sure that the other Python versions don't test the RUBI module.

asmeurer commented:
Also make it so that SymPy can still be imported and used in Python 3.6 if matchpy is not installed.

arihantparsoya committed:
fixed tests for python versions < 3.6

added rubi to sympy.integrals

fixed SqrtQ

fixed test_setup.py

Upabjojr commented on sympy/integrals/rubi/constraint.py:
```diff
@@ -0,0 +1,79 @@
+if matchpy:
+    Constraint, substitute = matchpy.Constraint, matchpy.substitute
+else:
+    raise ImportError('MatchPy could not be imported')
```
This should be avoided. We need to find a way to skip this module without raising an error, otherwise the tests will fail.

parsoyaarihant replied:
I have used `disabled = True` to stop running the test but it doesn't seem to be working in python2.7:

https://github.com/sympy/sympy/blob/9670ff89dd8bff7878c6d636ec4f493b87d092df/sympy/external/tests/test_sage.py#L27

Upabjojr replied:
We have the doctests problem as well. Doctests try to import every Python file.

parsoyaarihant replied:
I can use `sys.exit()`(as done [here](https://github.com/sympy/sympy/blob/9ab4041226f8f472d278f393b92469cd4926ecee/bin/mailmap_update.py#L13)) but that will stop the whole test itself.

Upabjojr replied:
The tests shouldn't be stopped.

Upabjojr commented on sympy/integrals/rubi/patterns.py:
```diff
@@ -0,0 +1,2424 @@
+matchpy = import_module("matchpy")
+
+if matchpy:
+    Wildcard, Pattern, ReplacementRule, ManyToOneReplacer = matchpy.Wildcard, matchpy.Pattern, matchpy.ReplacementRule, matchpy.ManyToOneReplacer
```
is this line making the code quality tests fail?

parsoyaarihant replied:
No, there is an old test which is failing code quality test for `patterns.py` file (only in python 2.7):

https://travis-ci.org/sympy/sympy/jobs/254914808

parsoyaarihant replied:
I am inspecting it now.

asmeurer commented:
As noted on gitter, the rubi modules should be importable in every Python version. Matchpy should only be imported when the rubi functions are called, so that imports don't raise ImportError. 

For the doctest, you should add support for checking Python versions to @doctest_depends_on and use that.

arihantparsoya committed:
fixed bin/test and bin/doctest for python<3.6

fixed code quality tests

parsoyaarihant commented:
Tests have passed

Upabjojr commented:
@parsoyaarihant What would you prefer: keep this PR open and work on it or merge it?

arihantparsoya committed:
added all algebriac rules using Wildcard.optional

parsoyaarihant commented:
I would prefer keeping this PR open. After Manuel implements code generation, the code will need to be changed again.

Upabjojr commented on sympy/integrals/rubi/symbol.py:
```diff
@@ -0,0 +1,20 @@
+class VariableSymbol(Symbol):
+    pass
+
+class Integer(Symbol):
```
There could be a potential name conflict with SymPy's Integer class. I generally suggest not to give the same name to two different classes in SymPy.

parsoyaarihant replied:
Shall I rename it to `Number`?

Upabjojr replied:
Maybe `MatchpyInteger`?

parsoyaarihant replied:
okay. I will use `MatchpyInteger`

parsoyaarihant commented:
How should I arrange the test suit? Rubi test suit is divided by different types of expressions:
```python
        # Integrands of the form a

        [S(0), x, S(1), S(0)],
        [S(1), x, S(1), x],
        [S(5), x, S(1), S(5)*x],
        [ - S(2), x, S(1), - S(2)*x],

        # Integrands of the form x**m
        [x**m, x, S(1), x**(S(1) + m)/(S(1) + m)],
        [x**S(100), x, S(1), S(1)/S(101)*x**S(101)],
        [x**S(3), x, S(1), S(1)/S(4)*x**S(4)],

        # Integrands of the form x**(m/S(2))
        [x**(S(5)/S(2)), x, S(1), S(2)/S(7)*x**(S(7)/S(2))],
        [x**(S(3)/S(2)), x, S(1), S(2)/S(5)*x**(S(5)/S(2))],
        [x**(S(1)/S(2)), x, S(1), S(2)/S(3)*x**(S(3)/S(2))],

        # Integrands of the form x**(m/S(3))
        [x**(S(5)/S(3)), x, S(1), S(3)/S(8)*x**(S(8)/S(3))],
        [x**(S(4)/S(3)), x, S(1), S(3)/S(7)*x**(S(7)/S(3))],
        [x**(S(2)/S(3)), x, S(1), S(3)/S(5)*x**(S(5)/S(3))],
        ...

```
Shall I divide these into indivdual `test_` functions(this could take lot of time since there are many types of expressions)?

arihantparsoya committed:
added rubi_test

Abdullahjavednesar committed:
rubi to integrals

arihantparsoya committed:
added all algebriac rules

arihantparsoya committed:
added all variablenames for algenriac rules

arihantparsoya committed:
added more utility functions and operations to Rubi

arihantparsoya committed:
removed statistics module

parsoyaarihant commented:
I have added all the algebriac rules, but adding them has made the matcher really slow. Even trivial expressions are also not getting matched for long time and I had to close the program locally.

Because of this, I am stopping the current travis test until this issue gets fixed.

arihantparsoya committed:
changed Integer to matchpyInteger

arihantparsoya committed:
Added RectifyTangent and RectifyCotangent

arihantparsoya committed:
added algebriac 1.2 tests

Upabjojr commented:
@parsoyaarihant as a low priority issue, there are probably classes inside this code that have the same name as SymPy classes. This could cause some problems for people who use the auto-import (i.e. an editor that fills the imports automatically). I would suggest to put a prefix so that we can be sure that there are no name conflicts.

parsoyaarihant commented:
I have renamed `Integer` to `matchpyInteger`. `Symbol` is inbuilt class in MatchPy so I dont think I can do anything about it.

parsoyaarihant commented:
@Upabjojr , I am trying to use SymPy objects in MatchPy. The class naming should not be there after that.

arihantparsoya committed:
Renamed `Integer` to `matchpyInteger` in patterns.py

arihantparsoya committed:
updated test_1_2.py

Abdullahjavednesar committed:
added functions

Abdullahjavednesar committed:
resolved conflicts

arihantparsoya committed:
Removed dependency on matchpy Operations and matchpy-sympy converters

Abdullahjavednesar committed:
new updates

Abdullahjavednesar committed:
removed test sets

Abdullahjavednesar committed:
removed test_rubi_test_suit.py

arihantparsoya committed:
Merge pull request #13 from Abdullahjavednesar/ajn_rubi9

added last set

arihantparsoya committed:
added tests in test_utility_function

Abdullahjavednesar committed:
tests fixed

Abdullahjavednesar committed:
Merge branch 'rubi4' of https://github.com/parsoyaarihant/sympy into ajn_rubi10

arihantparsoya committed:
added some utility functions and completed 1.2 tests suit

Abdullahjavednesar committed:
fixed TrigToExp test

Abdullahjavednesar committed:
updated

Abdullahjavednesar committed:
tests for IntSum

arihantparsoya committed:
Merge pull request #14 from Abdullahjavednesar/ajn_rubi10

Thanks

arihantparsoya committed:
fixed code_quality.py

arihantparsoya committed:
fixed bugs in utility functions

arihantparsoya committed:
added test_1_4.py

Abdullahjavednesar committed:
tests for quadratic\2.2

Abdullahjavednesar committed:
updated

arihantparsoya committed:
fixed bug in SImpHelp

added changes in test_1_4.py

Abdullahjavednesar committed:
PowerVariableDegree test fixed

arihantparsoya committed:
aadded for algebraic rules 1.3

added test_1_3.py

arihantparsoya committed:
Fixed simplerQ

arihantparsoya committed:
fixed test_args.py and test suit

Upabjojr committed:
print Rational is a SymPy reconstructible way

arihantparsoya committed:
added TrigSimplify, SimpFixFactor, SimplifyAntiderivative

fixed doctest for python < 2.7

arihantparsoya committed:
added get_matching_rule_definition()

arihantparsoya committed:
renamed WC to matchpyWC

Previously, `WC` was used directly in patterns. I have defined another function `WC()` which returns `matchpyWC` object. Now, we don't have to define Wildcard everytime with additional parameters.

Upabjojr commented on sympy/integrals/rubi/utility_function.py:
```diff
@@ -2710,7 +2710,7 @@ def PolynomialTermQ(u, x):
     n = Wild('n', exclude=[x])
     Match = u.match(a*x**n)
-    if Match and IntegerQ(Match[n]) and Match[n]>0:
+    if Match and IntegerQ(Match[n]) and Greater(Match[n], S(0)):
```
does `Greater` behave exactly as Mathematica's comparison does?

parsoyaarihant replied:
No, `Greater` is implemented such that it returns `False` if the inequality is `Relational`.

parsoyaarihant replied:
Mathematica returns relational when it cannot evaluate an inequality:
```mathematica
In[7]:= Re[x] > 0
Out[7]= Re[x]>0
```

arihantparsoya committed:
Changed the rules into readable format

arihantparsoya committed:
Added parsetools

Abdullahjavednesar committed:
Merge branch 'rubi4' of https://github.com/parsoyaarihant/sympy into ajn_rubi11

arihantparsoya committed:
fixed Integers in patterns.py

There were some errors while matching constraints such as `SumSimplerQ(m, 1)`. Here `1` was interpreted as Python `Int` hence, there was errors.

I have encapsulated all integers inside `S()`

fixed test setup

fixed doctest and test for parsetools

Fixed doctest of parsetools and test suit errors

fixed random failing tests

arihantparsoya committed:
Created seperate file for different rules

Abdullahjavednesar committed:
added SubstFor

Abdullahjavednesar committed:
updated

Abdullahjavednesar committed:
removed quadratic tests

arihantparsoya committed:
added all rules into Rubi

Added all rules into rubi_intergate().

Some Mathematica Special functions are not yet defined, Hence I have not imported Special functions

There are some parsing issues with derivative rules. I will add derivative rules as soon as the parser is fixed.

parsoyaarihant commented on sympy/integrals/rubi/utility_function.py:
```diff
@@ -0,0 +1,6259 @@
+        if len(match) == len(keys):
+            a, b, m, A, B, c, d = tuple([match[i] for i in keys])
+            if PositiveIntegerQ(m):
+            
```
@Abdullahjavednesar , complete these functions

Abdullahjavednesar replied:
@parsoyaarihant, you had implemented `ExpandIntegrand`, what's incomplete here?

parsoyaarihant replied:
`ElementaryFunctionQ`, `FunctionOfTrigOfLinearQ`

Abdullahjavednesar committed:
specialfunctions

Abdullahjavednesar committed:
update

arihantparsoya committed:
Added documentation in __init__.py file

Added documentation on how to use rubi module and TODO list.

commented out some rules in rubi.py since loading all the rules takes too much time for travis.

Abdullahjavednesar committed:
added error functions

Abdullahjavednesar committed:
Merge branch 'rubi4' of https://github.com/parsoyaarihant/sympy into ajn_rubi12

arihantparsoya committed:
Added documentation in functions

arihantparsoya committed:
Merge pull request #16 from Abdullahjavednesar/ajn_rubi12

Added SubstFor

certik commented:
@Upabjojr, @Abdullahjavednesar what needs to be done to finish this? Let's get it merged before the GSoC is over, i.e. this week?

Upabjojr commented:
> what needs to be done to finish this?

I guess we just need to add some `@XFAIL` or skip the tests that are failing. Some problems are in MatchPy and depend on that project.

certik commented:
@Abdullahjavednesar can you please skip the failing tests? Also, can you list the problems with MatchPy? Let's figure out what should be done.

parsoyaarihant commented:
@certik , there is a [bug](https://github.com/HPAC/matchpy/issues/26) in MatchPy because of which the tests are not passing.

Shall I skip the test anyway for now?

certik commented:
How many tests are we talking about? If it is just a few, then yes. If it is all of them, then let's see how to fix it.

parsoyaarihant commented:
Its only few(around 1-2). I will update this branch with the changes today.

Upabjojr commented:
Otherwise just put those tests in a separate location. We need to merge this and for that we need the tests to pass.

Abdullahjavednesar commented:
@parsoyaarihant, are you working on the failing tests?
@certik, Arihant has pointed the issue, once it is fixed I think we'll pass all the tests.

certik commented:
We need to merge this now. Let's skip the tests.

Sent from my mobile phone.

On Sep 1, 2017 6:18 AM, "Abdullah Javed Nesar" <notifications@github.com>
wrote:

> @parsoyaarihant <https://github.com/parsoyaarihant>, are you working on
> the failing tests?
> @certik <https://github.com/certik>, Arihant has pointed the issue, once
> it is fixed I think we'll pass all the tests.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/sympy/sympy/pull/12978#issuecomment-326565844>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AABQWDVMCBysUov3FOT-UlY4aNQCNG9jks5sd_YZgaJpZM4OZU6v>
> .
>

Upabjojr commented:
I agree, we haven't included into master any coffee of your work so far, it's time to merge, whatever the state of the code.

parsoyaarihant commented:
Okay, I am working on it right now.

arihantparsoya committed:
Added new rules by adding seperate With() functions

Previously, we were getting recursion errors due to `With`. I have defiend `With`outside the functions. This helps in preventing recursion errors.

Added error functions in matchpy matcher.

Removed additional parameters of gamma.

arihantparsoya committed:
Added @XFAIL decorator

certik commented:
I ran tests with Python 2.7:
```
$ bin/test sympy/integrals/
============================= test process starts ==============================
executable:         /home/certik/ext/miniconda2/envs/jup/bin/python  (2.7.12-final-0) [CPython]
architecture:       64-bit
cache:              yes
ground types:       python 
random seed:        84565124
hash randomization: on (PYTHONHASHSEED=1447910273)

sympy/integrals/tests/test_deltafunctions.py[2] ..                          [OK]
sympy/integrals/tests/test_failing_integrals.py[5] wwwww                    [OK]
sympy/integrals/tests/test_heurisch.py[25] .............www.........        [OK]
sympy/integrals/tests/test_integrals.py[107] .......www.........................
........................................................................    [OK]
sympy/integrals/tests/test_intpoly.py[3] ...                                [OK]
sympy/integrals/tests/test_lineintegrals.py[1] .                            [OK]
sympy/integrals/tests/test_manual.py[22] ......................             [OK]
sympy/integrals/tests/test_meijerint.py[25] ......www................       [OK]
sympy/integrals/tests/test_prde.py[15] ...............                      [OK]
sympy/integrals/tests/test_quadrature.py[16] ................               [OK]
sympy/integrals/tests/test_rationaltools.py[8] ........                     [OK]
sympy/integrals/tests/test_rde.py[9] .........                              [OK]
sympy/integrals/tests/test_risch.py[34] ..................................  [OK]
sympy/integrals/tests/test_singularityfunctions.py[1] .                     [OK]
sympy/integrals/tests/test_transforms.py[18] ...wwwww..........             [OK]
sympy/integrals/tests/test_trigonometry.py[4] ....                          [OK]
________________________________ slowest tests _________________________________
test_fourier_transform - Took 10.020 seconds
test_inverse_laplace_transform - Took 11.187 seconds
test_issue_10847 - Took 11.497 seconds
test_messy - Took 12.258 seconds
test_polytope_integrate - Took 16.016 seconds
test_expint - Took 21.027 seconds
========== tests finished: 276 passed, 19 skipped, in 329.11 seconds ===========
```
and I am now running it with Python 3.6 + matchpy. It is extremely slow....

It's probably due to the following issue:
```
In [1]: %time from sympy.integrals.rubi.rubi import rubi_integrate
CPU times: user 2min 8s, sys: 160 ms, total: 2min 9s
Wall time: 2min 9s
```
It takes over 2 minutes to import `rubi_integrate`. That will have to be fixed.

Otherwise it seems to work. However, it seems to be easily an order of magnitude slower than our current integrator, e.g.:
```
In [10]: %time rubi_integrate(x**6/(a + b*x)**2, x)
CPU times: user 4.26 s, sys: 0 ns, total: 4.26 s
Wall time: 4.25 s
Out[10]: 
        6           5                   4        3  2    2  3      4     5 
       a         6⋅a ⋅log(a + b⋅x)   5⋅a ⋅x   2⋅a ⋅x    a ⋅x    a⋅x     x  
- ──────────── - ───────────────── + ────── - ─────── + ───── - ──── + ────
   7                      7             6         5        4       3      2
  b ⋅(a + b⋅x)           b             b         b        b     2⋅b    5⋅b 

In [11]: %time integrate(x**6/(a + b*x)**2, x)
CPU times: user 456 ms, sys: 0 ns, total: 456 ms
Wall time: 458 ms
Out[11]: 
        6          5                   4        3  2    2  3      4     5 
       a        6⋅a ⋅log(a + b⋅x)   5⋅a ⋅x   2⋅a ⋅x    a ⋅x    a⋅x     x  
- ─────────── - ───────────────── + ────── - ─────── + ───── - ──── + ────
     7    8              7             6         5        4       3      2
  a⋅b  + b ⋅x           b             b         b        b     2⋅b    5⋅b 
```
Why is `rubi_integrate` 10x slower than `integrate`?

Why does importing `rubi_integrate` takes over 2 minutes?

certik commented:
So with Python 2.7 (no matchpy), the `sympy/integrals` take under 3 minutes on my machine. With Python 3.6 + matchpy, in over 45 minutes it only got this far:
```
$ bin/test sympy/integrals/
============================= test process starts ==============================
executable:         /home/certik/ext/miniconda2/envs/matchpy/bin/python  (3.6.2-final-0) [CPython]
architecture:       64-bit
cache:              yes
ground types:       python 
random seed:        729680
hash randomization: on (PYTHONHASHSEED=4108166084)

sympy/integrals/rubi/parsetools/tests/test_parse.py[1] .                    [OK]
sympy/integrals/rubi/tests/test_1_2.py[8] ..f...
```

Which test on Travis is testing matchpy? It seems it isn't tested, unless I missed it.

@Upabjojr, @parsoyaarihant, @Abdullahjavednesar, @asmeurer any ideas what we can do to move forward?

Upabjojr commented:
> It takes over 2 minutes to import rubi_integrate. That will have to be fixed.

We are well aware of this issue. We have discussed with @wheerd about the possibility of a code generator and that's probably the best solution to this problem. Unfortunately the generated code exceeds 20 MB and apparently that's a Python file size limit or something of a problem like that (if I have understood correctly).

We need to wait until the MatchPy developers solve this issue.

The RUBI tests are run on Python 3.6 only, because that's the only version that MatchPy supports.

Upabjojr commented:
> which finishes in under 3 minutes.

I guess tests for versions of Python other than 3.6 only test the utility functions of RUBI.

certik commented:
@Upabjojr I am actually -1 to merge as it is, because Travis does not test this PR at all.... More over, if you actually run the tests under matchpy, then they hang (!): #13239.

But since it's merged, we need to fix this asap.

certik commented:
I created a meta issue with all the problems that I encountered so far: #13243.

@Abdullahjavednesar, @parsoyaarihant please try to fix at least the testing issues.

asmeurer commented:
We can revert the merge if necessary (there is a "revert" button above that will do it).

certik commented:
I don't want to mess up the history even more. I think most people don't have python 3.6 *and* matchpy installed, and then they will never see the problems. However, @Abdullahjavednesar, @parsoyaarihant, let's try to fix this quickly please.

Upabjojr commented:
> I am actually -1 to merge as it is, because Travis does not test this PR at all.... 

Yes, it does: https://travis-ci.org/sympy/sympy/jobs/270968939

You should read the discussion on gitter, we have extensively discussed about the current problems with Python versions, loading time and MatchPy.

certik commented:
You are right. It is tested on Travis. That's a relief. It still hangs on
my computer, so that's a big problem, but since it works on Travis, it's
hopefully something that can be fixed. Perhaps I need to install gmpy, or
something.

In that case I am fine with the merge.

Sent from my mobile phone.

On Sep 1, 2017 11:27 PM, "Francesco Bonazzi" <notifications@github.com>
wrote:

I am actually -1 to merge as it is, because Travis does not test this PR at
all....

Yes, it does: https://travis-ci.org/sympy/sympy/jobs/270968939

You should read the discussion on gitter, we have extensively discussed
about the current problems with Python versions, loading time and MatchPy.

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub
<https://github.com/sympy/sympy/pull/12978#issuecomment-326723252>, or mute
the thread
<https://github.com/notifications/unsubscribe-auth/AABQWGeHftX7eya5MtY7wL7ZR9FeUPurks5seOc4gaJpZM4OZU6v>
.

parsoyaarihant commented:
> . It still hangs on my computer, so that's a big problem,

It will be fixed by #13249

certik commented:
@parsoyaarihant thanks for the work, indeed it did.

I think we are in good shape regarding this PR, it is tested on Travis and the tests pass on my machine reasonably quickly. So all is good.

Thanks @parsoyaarihant, @Abdullahjavednesar and @Upabjojr for your work on this!
